File: sim/runaway_sim/gpx.py
"""GPX 파싱 → 1Hz 트랙. 입력 단계라 float 사용. 코어로 넘길 때 cm 정수로 변환."""
import xml.etree.ElementTree as ET
from datetime import datetime, timezone
from .geo import haversine_m

GATE_ACC_M = 30.0      # 이 값 초과 정확도는 폐기(직전 위치 유지)
GATE_SPEED_MS = 8.0    # 8 m/s 초과 이동은 폐기
DROPOUT_S = 20         # 이 시간 이상 점이 없으면 신호 소실로 처리


def _parse_time(s):
    s = s.strip().replace('Z', '+00:00')
    return datetime.fromisoformat(s).astimezone(timezone.utc)


def parse_gpx(path):
    """[(t_s, lat, lon, acc_m)] 원본 점. acc는 extensions의 accuracy 또는 hdop*5, 없으면 8."""
    tree = ET.parse(path)
    pts = []
    for el in tree.iter():
        if el.tag.endswith('trkpt'):
            lat, lon = float(el.get('lat')), float(el.get('lon'))
            t = None
            acc = None
            for c in el.iter():
                tag = c.tag.split('}')[-1]
                if tag == 'time' and c.text:
                    t = _parse_time(c.text)
                elif tag == 'accuracy' and c.text:
                    acc = float(c.text)
                elif tag == 'hdop' and c.text and acc is None:
                    acc = float(c.text) * 5.0
            if t is None:
                continue
            pts.append((t, lat, lon, 8.0 if acc is None else acc))
    if not pts:
        raise ValueError('trkpt 없음: ' + path)
    t0 = pts[0][0]
    return [((p[0] - t0).total_seconds(), p[1], p[2], p[3]) for p in pts]


def resample_1hz(raw):
    """선형 보간으로 1Hz. 20초 이상 공백은 dropout: 위치 유지, acc 99."""
    out = []
    i = 0
    t_end = int(raw[-1][0])
    for t in range(0, t_end + 1):
        while i + 1 < len(raw) and raw[i + 1][0] <= t:
            i += 1
        a = raw[i]
        if i + 1 < len(raw):
            b = raw[i + 1]
            if b[0] - a[0] > DROPOUT_S and t > a[0]:
                out.append((t, a[1], a[2], 99.0))
                continue
            f = 0.0 if b[0] == a[0] else (t - a[0]) / (b[0] - a[0])
            f = max(0.0, min(1.0, f))
            out.append((t, a[1] + (b[1] - a[1]) * f, a[2] + (b[2] - a[2]) * f, a[3] + (b[3] - a[3]) * f))
        else:
            out.append((t, a[1], a[2], a[3]))
    return out


SMOOTH_ALPHA = 0.35    # EMA. 백색 잡음 억제
JUMP_RESET_N = 3       # 연속 3회 점프면 새 앵커로 인정(실제 순간이동 복구)


def to_track(pts_1hz):
    """스무딩 + 게이트 필터 후 [(t, dist_cm 누적, speed_cms, acc_m)]. 코어 입력 형식."""
    track = []
    dist = 0.0
    slat = slon = plat = plon = float('nan')
    ignored = 0
    for (t, lat, lon, acc) in pts_1hz:
        if acc > GATE_ACC_M:
            track.append((t, int(round(dist * 100)), 0, acc))
            continue                      # 저정확도: 위치·거리 유지
        if slat != slat:                  # 게이트를 통과한 첫 점만 평활화 시드
            slat, slon = lat, lon
            plat, plon = lat, lon
        slat = SMOOTH_ALPHA * lat + (1 - SMOOTH_ALPHA) * slat
        slon = SMOOTH_ALPHA * lon + (1 - SMOOTH_ALPHA) * slon
        step = haversine_m(plat, plon, slat, slon)
        if step > GATE_SPEED_MS:
            ignored += 1
            if ignored >= JUMP_RESET_N:
                plat, plon = slat, slon   # 앵커 리셋, 거리는 더하지 않음
                ignored = 0
            step = 0.0
        else:
            ignored = 0
            plat, plon = slat, slon
        dist += step
        track.append((t, int(round(dist * 100)), int(round(step * 100)), acc))
    return track


def load_track(path):
    return to_track(resample_1hz(parse_gpx(path)))

File: core/src/main/kotlin/kr/runaway/core/Gpx.kt
package kr.runaway.core

import kotlin.math.asin
import kotlin.math.cos
import kotlin.math.round
import kotlin.math.sin
import kotlin.math.sqrt

/** GPX 파싱, 1Hz 리샘플, 스무딩·게이트 필터. sim/runaway_sim/gpx.py와 동일 연산. */
object Gpx {
    const val GATE_ACC_M = 30.0
    const val GATE_SPEED_MS = 8.0
    const val DROPOUT_S = 20
    const val SMOOTH_ALPHA = 0.35
    const val JUMP_RESET_N = 3
    private const val EARTH_R = 6371000.0

    data class Pt(val t: Double, val lat: Double, val lon: Double, val acc: Double)

    fun haversineM(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val p1 = Math.toRadians(lat1); val p2 = Math.toRadians(lat2)
        val dp = p2 - p1; val dl = Math.toRadians(lon2 - lon1)
        val a = sin(dp / 2) * sin(dp / 2) + cos(p1) * cos(p2) * sin(dl / 2) * sin(dl / 2)
        return 2 * EARTH_R * asin(sqrt(a))
    }

    private val trkpt = Regex("""<trkpt[^>]*lat="([^"]+)"[^>]*lon="([^"]+)"[^>]*>(.*?)</trkpt>""", RegexOption.DOT_MATCHES_ALL)
    private val timeRe = Regex("""<time>([^<]+)</time>""")
    private val accRe = Regex("""<accuracy>([^<]+)</accuracy>""")
    private val hdopRe = Regex("""<hdop>([^<]+)</hdop>""")

    /** 원본 점. t는 첫 점 기준 초. acc는 accuracy 또는 hdop*5, 없으면 8. */
    fun parse(xml: String): List<Pt> {
        val out = ArrayList<Pt>()
        var t0: Long? = null
        for (m in trkpt.findAll(xml)) {
            val body = m.groupValues[3]
            val tm = timeRe.find(body) ?: continue
            val epochMs = java.time.Instant.parse(tm.groupValues[1].trim()).toEpochMilli()
            if (t0 == null) t0 = epochMs
            val acc = accRe.find(body)?.groupValues?.get(1)?.toDouble()
                ?: hdopRe.find(body)?.groupValues?.get(1)?.toDouble()?.times(5.0) ?: 8.0
            out.add(Pt((epochMs - t0) / 1000.0, m.groupValues[1].toDouble(), m.groupValues[2].toDouble(), acc))
        }
        require(out.isNotEmpty()) { "trkpt 없음" }
        return out
    }

    fun resample1Hz(raw: List<Pt>): List<Pt> {
        val out = ArrayList<Pt>()
        var i = 0
        val tEnd = raw.last().t.toInt()
        for (t in 0..tEnd) {
            while (i + 1 < raw.size && raw[i + 1].t <= t) i++
            val a = raw[i]
            if (i + 1 < raw.size) {
                val b = raw[i + 1]
                if (b.t - a.t > DROPOUT_S && t > a.t) { out.add(Pt(t.toDouble(), a.lat, a.lon, 99.0)); continue }
                var f = if (b.t == a.t) 0.0 else (t - a.t) / (b.t - a.t)
                f = f.coerceIn(0.0, 1.0)
                out.add(Pt(t.toDouble(), a.lat + (b.lat - a.lat) * f, a.lon + (b.lon - a.lon) * f, a.acc + (b.acc - a.acc) * f))
            } else out.add(Pt(t.toDouble(), a.lat, a.lon, a.acc))
        }
        return out
    }

    /** 스트리밍 트랙 빌더. 리플레이와 실시간 위치 모두 이걸 통해 코어 입력을 만든다. */
    class TrackBuilder {
        private var dist = 0.0
        private var slat = Double.NaN; private var slon = Double.NaN
        private var plat = Double.NaN; private var plon = Double.NaN
        private var ignored = 0
        var lastSmoothLat = Double.NaN; private set
        var lastSmoothLon = Double.NaN; private set

        fun push(t: Int, lat: Double, lon: Double, acc: Double): TrackRow {
            // 정확도 게이트를 먼저. 나쁜 점이 평활화 시드가 되면 좌표가 통째로 날아간다
            if (acc > GATE_ACC_M) return TrackRow(t, round(dist * 100).toInt(), 0, acc)
            if (slat.isNaN()) { slat = lat; slon = lon; plat = lat; plon = lon }
            slat = SMOOTH_ALPHA * lat + (1 - SMOOTH_ALPHA) * slat
            slon = SMOOTH_ALPHA * lon + (1 - SMOOTH_ALPHA) * slon
            lastSmoothLat = slat; lastSmoothLon = slon
            var step = haversineM(plat, plon, slat, slon)
            if (step > GATE_SPEED_MS) {
                ignored += 1
                if (ignored >= JUMP_RESET_N) { plat = slat; plon = slon; ignored = 0 }
                step = 0.0
            } else { ignored = 0; plat = slat; plon = slon }
            dist += step
            return TrackRow(t, round(dist * 100).toInt(), round(step * 100).toInt(), acc)
        }
    }

    fun toTrack(pts: List<Pt>): List<TrackRow> {
        val b = TrackBuilder()
        return pts.map { b.push(it.t.toInt(), it.lat, it.lon, it.acc) }
    }

    fun loadTrack(xml: String): List<TrackRow> = toTrack(resample1Hz(parse(xml)))
}
